### ISSUE TITLE
Python as a virtual dependency

### ISSUE THREAD

**Issue by mkuron** [spack/spack#7966]
I'm creating a separate issue for this request that was previously discussed at https://github.com/spack/spack/issues/2215#issuecomment-346464285 with @alalazo.

There are multiple Python packages available in Spack (_python_, _miniconda2_, _miniconda3_); an Anaconda package has been requested (#2738, #2741), and some of the alternative interpreters (e.g. Jython, PyPy) might also be packaged at some point (even though they only support a subset of Python libraries). Since all these packages provide a Python interpreter and libraries, there should logically be a `Python` virtual package and the existing `Python` package should be renamed to `CPython`.

This has two advantages:
- Python can be used in lmod hierarchies. (This is why we discussed it in #2215).
- Python libraries (e.g. _py-numpy_) can now be installed for one of the other providers.

@citibeth voiced her concerns about making such a change as the `Python` package is a bit special and the change might break other things. However, with #7955 this might become easier to implement.

I already tried to implement this change, but ran into some issues (https://github.com/spack/spack/issues/2215#issuecomment-349661611). @alalazo also wanted to give it a try.

**Comment by citibeth** [comment]
@mkuron As I'm sure I've said before, I think this is a rather speculative proposal; and so is #7955.  Maybe we should try the two together, and see how / if it works?  That would put us all in the best position to decide whether (a) we want to move away from special-purpose Virtual Deps, and (b) whether we want to move toward `python` as a virtual dep.

If we are to go ahead with a working prototype, I would want minimal buy-in from @tgamblin that these ideas are not completely out of the question.  With a positive (non-negative) grunt from @tgamblin I'd be happy to see if I can hash out an explicit `VirtualPackage` class on a new shared PR.  You would add a new virtualized Python package, rename the old Python to `cpython`, and add at least one more Python provider.  We then try it out, see what breaks and evaluate?

Your thoughts?

**Comment by mkuron** [comment]
Sure, sounds reasonable.

**Comment by adamjstewart** [comment]
I definitely agree that we need a `python` virtual dependency.

> One cannot currently extend virtual packages as the virtual dependencies appear to not have been concretized by the time the extends is performed.

This seems like the only hurdle. Not sure where to start on this one.

**Comment by mkuron** [comment]
> This seems like the only hurdle. Not sure where to start on this one.

I guess that issue would be gone with #7955.

**Comment by citibeth** [comment]
I agree.  Also, `extends` and activation are not strictly necessary, and
they break the overall Spack model.

(Spack implicitly assumes that two installed packages of the same hash are
the same.  Updating files in the package post-installation breaks this
assumption.  Spack activation is one such way to break it, meaning that
some things that should work with Spack don't when using package
activation.  Instead of activating, loading all the modules to be activated
is a more Spack-like alternative.)

**Comment by adamjstewart** [comment]
> Also, `extends` and activation are not strictly necessary, and they break the overall Spack model.

Nevertheless, they are core components of Spack used by many who don't want to have to load a dozen dependencies just to use a single Python package. We don't want to break this behavior.

**Comment by adamjstewart** [comment]
>> This seems like the only hurdle. Not sure where to start on this one.
>
> I guess that issue would be gone with #7955.

I would guess again. #7955 got 3 😕 reactions from core developers. I don't expect it to happen unless it is more fleshed out. It sounds like a nice idea, until you realize how Spack works.

**Comment by citibeth** [comment]
>
>
> I would guess again. #7955 <https://github.com/spack/spack/issues/7955>
> got 3 😕 reactions from core developers.
>

I agree, that's not a a good sign to proceed.  I'd be interested in
learning more about the reasons behind the negative reactions.  But as it
stands, I would not put any actual work into prototyping #7955.

**Comment by mkuron** [comment]
Ok, then maybe we should try to fix this one independently of #7955:

> One cannot currently extend virtual packages as the virtual dependencies appear to not have been concretized by the time the `extends` is performed.

**Comment by adamjstewart** [comment]
> I'd be interested in learning more about the reasons behind the negative reactions.

Virtual dependencies are one of the core tenants of Spack that have been there since day 1. I believe they are a large part of the original Spack publication. It's not that they're flawless, it's more that they are very entrenched in what it means to be "Spack". There is definitely more than one way to skin a cat, but if you want to convince the core devs to change this, it needs to be a lot more well thought out with a better argument for why it's necessary.

**Comment by citibeth** [comment]
@tgamblin wrote:
> vdeps are part of the core spack model.
@adamjstewart wrote:
> Virtual dependencies are one of the core tenants of Spack that have been there since day 1.

Other good reasons have been cited.  But I feel compelled to point out that this is NOT a good reason.  Compilers as an attribute of specs has also been part of Spack from day 1.  But now that we have build dependencies, there is talk of removing compilers as a special case and turning them into build dependencies.

In any case, I think the case on #7955 is settled for now.  I'd just like virtual deps to work better.  As for Python... given the continued buggy state of virtual deps and the centrality of Python, I'm afraid that turning Python into a virtual dep could create a lot of long-standing headaches for many people.  If we could get virtual dep bugs cleaned up, I'd feel differently about the idea of using them for Python.

Of course... someone could just try to do it and see what happens.  Maybe it's not so bad.

**Comment by tgamblin** [comment]
Can you guys tag the vdep issues if they’re not tagged already?  Would like to take a look at all of them.

**Comment by mrmundt** [comment]
As per the Spack tutorial today, there is still an interest in this. We would appreciate this extension to allow these to be added to a spack stack.

**Comment by mrmundt** [comment]
Is there any ETA/update on this?

**Comment by adamjstewart** [comment]
#17475 gives a pretty good summary of what needs to be fixed in order to support this feature. I haven't heard any talk about anyone trying to tackle this in the near future. @scheibelp @becker33?
